test/e2e/prometheus_test.go

```diff
@@ -92,3 +97,223 @@ func TestPrometheusAlertmanagerAntiAffinity(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// setup a self-signed ca and store the artifacts in a secret
+	tlsSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "selfsigned-mtls-bundle",
+			Namespace: f.Ns,
+		},
+		Data: map[string][]byte{
+			"client-cert-name": []byte("test-client"),
+			"serving-cert-url": []byte(deployedService.Spec.ClusterIP),
+		},
+	}
+	if err := createSelfSignedMTLSArtifacts(tlsSecret); err != nil {
+		t.Fatal(err)
+	}
+	if err := f.OperatorClient.CreateOrUpdateSecret(ctx, tlsSecret); err != nil {
+		t.Fatal(err)
+	}
+
+	// deploy remote write target
+	targetDeployment := fmt.Sprintf(`apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: instrumented-sample-app
+  namespace: openshift-monitoring
+  labels:
+    group: %[1]s
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      group: %[1]s
+  template:
+    metadata:
+      labels:
+        group: %[1]s
+    spec:
+      containers:
+      - name: example-app
+        args:
+        - --cert-path=/etc/certs
+        image: quay.io/coreos/instrumented-sample-app:0.2.0-bearer-mtls-1
+        imagePullPolicy: IfNotPresent
+        ports:
+        - name: web
+          containerPort: 8080
+        - name: mtls
+          containerPort: 8081
+        volumeMounts:
+        - mountPath: /etc/certs
+          name: certs
+      volumes:
+      - name: certs
+        secret:
+          secretName: selfsigned-mtls-bundle
+          items:
+          - key: server-ca.pem
+            path: cert.pem
+          - key: server.key
+            path: key.pem
+`, name)
+	rwTestDeployment, err := manifests.NewDeployment(bytes.NewReader([]byte(targetDeployment)))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := f.OperatorClient.CreateOrUpdateDeployment(ctx, rwTestDeployment); err != nil {
+		t.Fatal(err)
+	}
+	for _, scenario := range []struct {
+		name   string
+		port   string
+		rwSpec string
+	}{
+		// check remote write logs
+		{
+			name: "assert remote write to http works",
+			port: "8080",
+			rwSpec: `
+  - url: http://%s`,
+		},
+		{
+			name: "assert mtls remote write works",
+			port: "8081",
+			rwSpec: `
+  - url: https://%s
+    tlsConfig:
+      ca:
+        secret:
+          name: selfsigned-mtls-bundle
+          key: ca.crt
+      cert:
+        secret:
+          name: selfsigned-mtls-bundle
+          key: client.crt
+      keySecret:
+        name: selfsigned-mtls-bundle
+        key: client.key
+`,
+		},
+	} {
+		rw := fmt.Sprintf(scenario.rwSpec, deployedService.Spec.ClusterIP+":"+scenario.port)
+
+		cmoConfigMap := fmt.Sprintf(`prometheusK8s:
+  logLevel: debug
+  remoteWrite: %s
+`, rw)
+		if err := f.OperatorClient.CreateOrUpdateConfigMap(ctx, configMapWithData(t, cmoConfigMap)); err != nil {
+			t.Fatal(err)
+		}
+
+		t.Run(scenario.name, checkRemoteWrite(name, ctx))
+	}
+}
+
+func checkRemoteWrite(rwEndpointName string, ctx context.Context) func(*testing.T) {
+	return func(t *testing.T) {
+		remoteWriteCheckLogs(rwEndpointName, ctx, t)
```

Comment:
Do we still want to make assertions on the logs? I thought the idea was to use metrics instead since they are more stable than logs.

Reply:
I would argue yes. Errors in the logs will give a clearer indication to issues in the mTLS setup for example. Checking the metric give a clearer signal that the requests are actually happening.

Where do you expect instability regarding the logs or any downside with checking the logs?

Reply:
I don't expect instability per se, but I also do not think there is any guarantee that log messages would remain unchanged in future Prometheus versions. For example, someone might decide to reword a message for clarity, change the verbosity level of a message, change the format, add punctuation etc. Also, logs are usually meant to be read by humans. On the other hand, metrics are considered an API and are also intended for machines. Removing them is usually a breaking change which is documented in some release notes.

That being said, I think that in this particular test case it's fine to check the logs if using only metrics is not sufficient to verify correct behavior.